packages/components/src/DynamicDisclosure/DynamicDisclosure.tsx

```diff
@@ -108,51 +77,84 @@ export const useDynamicDisclosure = () => {
     } = {}
   ) => {
     const onClose = () => {
-      closeDisclosure();
-      reset();
+      setDisclosureStack(prev => {
+        const newStack = [...prev];
+        newStack[newStack.length - 1].isOpen = false;
+        return newStack;
+      });
+
+      if (disclosureStack.length <= 1) {
+        // Wait for the modal to close before removing it from the stack to keep original closing transition
+        setTimeout(() => {
+          setDisclosureStack(prev => prev.slice(0, -1));
+        }, TRANSITION_DURATION);
+      } else {
+        setDisclosureStack(prev => prev.slice(0, -1));
+      }
+
       void props.onClose?.();
-      setDisclosureProps(defaultProps);
     };
 
-    setMode(props.mode || "modal");
-    setDisclosureProps({ size: "md", ...props, onClose });
-    setDisclosureContent(content);
-    goToStep(content);
-    onOpen();
+    const newDisclosure = {
+      content,
+      props: { size: "md", ...props, onClose },
+      mode: props.mode || "modal",
+      isOpen: true,
+    };
+
+    setDisclosureStack(prev => [...prev, newDisclosure]);
     return Promise.resolve();
   };
 
   const renderContent = () => {
-    if (mode === "modal") {
-      return (
-        <Modal
-          autoFocus={false}
-          blockScrollOnMount={false}
-          closeOnOverlayClick={false}
-          isCentered
-          isOpen={isOpen}
-          motionPreset="slideInBottom"
-          {...disclosureProps}
-        >
-          <ModalOverlay />
-          <RemoveScroll enabled={isOpen}>{currentStep}</RemoveScroll>
-        </Modal>
-      );
-    } else {
-      return (
-        <Drawer isOpen={isOpen} placement="right" {...disclosureProps}>
-          <DrawerOverlay />
-          <RemoveScroll enabled={isOpen}>{currentStep}</RemoveScroll>
-        </Drawer>
-      );
-    }
+    if (disclosureStack.length === 0) return null;
+
+    return (
+      <>
+        {disclosureStack.map(({ content, props, mode, isOpen }, index) => {
+          const isTopmost = index === disclosureStack.length - 1;
+          const showOverlay =
+            (index > 0 && disclosureStack[index - 1].mode === "drawer") || index === 0;
+          const style = isTopmost ? {} : { display: "none" };
```

Comment:
I ain't sure if this logic would work tbh.
if we store the stack in a `ref` and keep the current index in the `state` then it should be fine because we don't make any copies of the existing components (e.g. no new component states).
but if we do copy the whole array then the state is going to be replaced with a fresh one once we click the goBack button.

having that given, I think it's not needed to render all of the modals all the time and hiding all but one, we can just render the one that is on that specific index stored in a `useState`. that way, we reuse the same object from memory and it should be fine.

though I might be wrong.

and, in any case, the key must not be `index`. since we don't really know what we store inside, `nanoid` as a key seems to be a good fit